applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,218 @@
+		reschedule_dfvs_retry_work(target_freq);
+
+	} else if (ret == -EBUSY) {
+		LOG_DBG("DVFS frequency change in progress");
+		reschedule_dfvs_retry_work(target_freq);
+
+	} else if (ret != 0) {
+		LOG_ERR("DVFS freq change returned with error: %d", ret);
+		module_set_state(MODULE_STATE_ERROR);
+	} else {
+		LOG_INF("Have requested %s frequency",
+			get_dvfs_frequency_setting_name(target_freq));
+		k_work_cancel_delayable(&DVFS_retry_data.DVFS_retry_work);
+		current_freq = target_freq;
+		DVFS_retry_data.dfvs_retries_cnt = 0;
+	}
+}
+
+static enum dvfs_frequency_setting check_required_frequency(void)
+{
+	enum dvfs_frequency_setting highest_requested_freq = DVFS_FREQ_LOW;
+
+	for (int i = 0; i < DVFS_STATE_COUNT; i++) {
+		if (dfvs_requests_state & BIT(i)) {
+			/* We have inversion in the freq DVFS_FREQ_LOW has the highest value */
+			if (highest_requested_freq > dvfs_freq_table[i]) {
+				highest_requested_freq = dvfs_freq_table[i];
+				if (highest_requested_freq == DVFS_FREQ_HIGH) {
+					break;
+				}
+			}
+		}
+	}
+
+	return highest_requested_freq;
+}
+
+static void update_dfvs_state(enum dvfs_state state, bool enable)
+{
+	if (enable) {
+		dfvs_requests_state |= BIT(state);
+		LOG_DBG("%s ACTIVE", get_dvfs_state_name(state));
+	} else {
+		dfvs_requests_state &= ~BIT(state);
+		LOG_DBG("%s NOT_ACTIVE", get_dvfs_state_name(state));
+	}
+}
+
+static void process_dvfs_states(enum dvfs_state state, bool turn_on)
+{
+	update_dfvs_state(state, turn_on);
+	enum dvfs_frequency_setting required_freq = check_required_frequency();
+
+	if (required_freq != current_freq) {
+		set_dvfs_freq(required_freq);
+	}
+}
+
+static void check_if_llpm_state(const struct ble_peer_conn_params_event *event)
+{
+	if (!event->updated) {
+		/* Ignore the connection parameters update request. */
+		return;
+	}
+
+	__ASSERT_NO_MSG(event->interval_min == event->interval_max);
+
+	process_dvfs_states(DVFS_STATE_LLPM_CONNECTED,
+			    event->interval_min & REG_CONN_INTERVAL_LLPM_MASK);
+}
+
+static void DVFS_retry_work_handler(struct k_work *work)
+{
+	LOG_DBG("Retrying to change DVFS frequency.");
+
+	set_dvfs_freq(DVFS_retry_data.target_freq);
+}
+
+static bool app_event_handler(const struct app_event_header *aeh)
+{
+	if (is_module_state_event(aeh)) {
+		const struct module_state_event *event =
+			cast_module_state_event(aeh);
+
+		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+
+			k_work_init_delayable(&DVFS_retry_data.DVFS_retry_work,
+					      DVFS_retry_work_handler);
+			process_dvfs_states(DVFS_STATE_INITIALIZING, false);
+		}
+
+		return false;
+	}
+
+	if (is_ble_peer_conn_params_event(aeh)) {
+		check_if_llpm_state(cast_ble_peer_conn_params_event(aeh));
+
+		return false;
+	}
+
+	if (is_ble_peer_event(aeh)) {
+		if (cast_ble_peer_event(aeh)->state == PEER_STATE_DISCONNECTED) {
+			process_dvfs_states(DVFS_STATE_LLPM_CONNECTED, false);
+		}
+
+		return false;
+	}
+
+	if (is_usb_state_event(aeh)) {
```

Comment:
`if (IS_ENABLED(CONFIG_DESKTOP_USB_ENABLE) && is_usb_state_event(aeh))`. Otherwise you would see linker errors related to undefined event when building with USB disabled.